pkg/worker/log/watcher_linux.go

```diff
@@ -0,0 +1,110 @@
+package log
+
+import (
+	"context"
+	"log"
+	"syscall"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+)
+
+type fileLinuxEvent struct {
+	mask uint32
+}
+
+// Modified returns true if the event is syscall.IN_MODIFY.
+func (e *fileLinuxEvent) Modified() bool {
+	return e.mask&syscall.IN_MODIFY == syscall.IN_MODIFY
+}
+
+// Deleted returns true if the event is syscall.IN_DELETE
+func (e *fileLinuxEvent) Deleted() bool {
+	return e.mask&syscall.IN_DELETE == syscall.IN_DELETE
+}
+
+// wrapLinuxEvent create a new File event given an event mask.
+func wrapLinuxEvent(mask uint32) FileEvent {
+	return &fileLinuxEvent{
+		mask: mask,
+	}
+}
+
+// linuxWatcher Watcher implementation for Linux
+type linuxWatcher struct{}
+
+// NewWatcher instance
+func NewWatcher() Watcher {
+	return &linuxWatcher{}
+}
+
+// Watcher watch file system events and send them throught a channel.
+// See https://man7.org/linux/man-pages/man7/inotify.7.html
+func (e *linuxWatcher) Watch(ctx context.Context, path string, mask uint32) (eventchan chan FileEvent, err error) {
+	filed, err := syscall.InotifyInit()
+	if filed == -1 || err != nil {
+		return
+	}
+	// watching for file system events given a file path
+	watched, err := unix.InotifyAddWatch(filed, path, mask)
+	if err != nil {
+		fderr := unix.Close(filed)
+		if fderr != nil {
+			log.Printf("Fail to close file descriptor: %v", err)
+		}
+		return
+	}
+	eventchan = make(chan FileEvent)
+	go func() {
+		// first, remove watched from the watch list (the order is importante here)
+		// close file watcher, filedescriptor and channel
+		defer func() {
+			success, err := unix.InotifyRmWatch(filed, uint32(watched))
+			if success == -1 || err != nil {
+				log.Printf("Fail to remove watch: %v", err)
+			}
+			err = unix.Close(watched)
+			if err != nil {
+				log.Printf("Fail to close watch: %v", err)
+			}
+			err = unix.Close(filed)
+			if err != nil {
+				log.Printf("Fail to close file descriptor: %v", err)
+			}
+			close(eventchan)
+		}()
+		// buffer to store events sent by OS
+		var buf [unix.SizeofInotifyEvent * 4096]byte
+		for {
+			// read events from file descriptor and
+			// fill the event buffer
+			nbytes, err := unix.Read(filed, buf[:])
+			if err != nil {
+				nbytes = 0
+				log.Printf("Fail to read events: %v", err)
+				continue
+			}
+			// iterate over all buffered events and send them through the
+			// event channel
+			var offset uint32
+			for offset <= uint32(nbytes-unix.SizeofInotifyEvent) {
```

Comment:
is this more accurate?
```suggestion
			for offset < uint32(nbytes) {
```